redis.go

```diff
@@ -0,0 +1,72 @@
+package iprepd
+
+import (
+	"math/rand"
+	"time"
+
+	"github.com/go-redis/redis"
+	log "github.com/sirupsen/logrus"
+)
+
+type redisLink struct {
+	master      *redis.Client
+	readClients []*redis.Client
+}
+
+func (r *redisLink) del(k ...string) *redis.IntCmd {
+	return r.master.Del(k...)
+}
+
+func (r *redisLink) flushAll() *redis.StatusCmd {
+	return r.master.FlushAll()
+}
+
+func (r *redisLink) get(k string) (ret []byte, err error) {
+	readRot := make([]*redis.Client, len(r.readClients))
+	p := rand.Perm(len(r.readClients))
+	for a, b := range p {
+		readRot[b] = r.readClients[a]
+	}
+	for i := range readRot {
+		ret, err = readRot[i].Get(k).Bytes()
+		if err == nil || (err != nil && err == redis.Nil) {
+			return
+		}
+		log.Error(err.Error())
+	}
+	// None of the read clients could satisfy the request, return the last
+	// error we have seen
+	return
+}
+
+func (r *redisLink) ping() *redis.StatusCmd {
+	return r.master.Ping()
+}
+
+func (r *redisLink) set(k string, v interface{}, e time.Duration) *redis.StatusCmd {
+	return r.master.Set(k, v, e)
+}
+
+func newRedisLink(cfg serverCfg) (ret redisLink, err error) {
+	ret.master = redis.NewClient(&redis.Options{
+		Addr:         cfg.Redis.Addr,
+		DB:           0,
+		ReadTimeout:  time.Millisecond * time.Duration(cfg.Redis.ReadTimeout),
+		WriteTimeout: time.Millisecond * time.Duration(cfg.Redis.WriteTimeout),
+		DialTimeout:  time.Millisecond * time.Duration(cfg.Redis.DialTimeout),
+	})
+	_, err = ret.ping().Result()
+	ret.readClients = make([]*redis.Client, 0)
+	for _, x := range cfg.Redis.Replicas {
+		y := redis.NewClient(&redis.Options{
+			Addr:        x,
+			DB:          0,
+			ReadTimeout: time.Millisecond * time.Duration(cfg.Redis.ReadTimeout),
+			DialTimeout: time.Millisecond * time.Duration(cfg.Redis.DialTimeout),
+		})
+		ret.readClients = append(ret.readClients, y)
+	}
+	// Also use the master for reads
+	ret.readClients = append(ret.readClients, ret.master)
```

Comment:
👍

Reply:
So turns out that the way the cloudformation output / export works for read replica addresses, it appends the primary _for_ you. It might be worth uniq'ing the different read addresses.

Reply:
@milescrabill ok; yeah let's just verify uniqueness on the hosts on config load that seems better for usability outside AWS